p2p/memp2p/memp2p.go

```diff
@@ -0,0 +1,356 @@
+
+	if !found {
+		return p2p.ErrNilTopic
+	}
+
+	if validator != nil {
+		return p2p.ErrTopicValidatorOperationNotSupported
+	}
+
+	messenger.Topics[topic] = handler
+	return nil
+}
+
+// UnregisterMessageProcessor unsets the message processor for the given topic
+// (sets it to nil).
+func (messenger *MemP2PMessenger) UnregisterMessageProcessor(topic string) error {
+	fmt.Printf("memp2p:UnregisterMessageProcessor\n")
+	messenger.TopicsMutex.Lock()
+	defer messenger.TopicsMutex.Unlock()
+	validator, found := messenger.Topics[topic]
+
+	if !found {
+		return p2p.ErrNilTopic
+	}
+
+	if validator == nil {
+		return p2p.ErrTopicValidatorOperationNotSupported
+	}
+
+	messenger.Topics[topic] = nil
+	return nil
+}
+
+// OutgoingChannelLoadBalancer does nothing, as it is not applicable to the in-memory network.
+func (messenger *MemP2PMessenger) OutgoingChannelLoadBalancer() p2p.ChannelLoadBalancer {
+	return nil
+}
+
+// BroadcastOnChannelBlocking sends the message to all peers in the network. It
+// calls parametricBroadcast() with async=false, which means that peers will
+// have their ReceiveMessage() function called synchronously. The call
+// to parametricBroadcast() is done synchronously as well. This function should
+// be called as a go-routine.
+func (messenger *MemP2PMessenger) BroadcastOnChannelBlocking(channel string, topic string, buff []byte) {
+	fmt.Printf("memp2p:BroadcastOnChannelBlocking\n")
+	messenger.parametricBroadcast(topic, buff, false)
+}
+
+// BroadcastOnChannel sends the message to all peers in the network. It calls
+// parametricBroadcast() with async=false, which means that peers will have
+// their ReceiveMessage() function called synchronously. The call to
+// parametricBroadcast() is done as a go-routine, which means this function is,
+// in fact, non-blocking, but it is identical with BroadcastOnChannelBlocking()
+// in all other regards.
+func (messenger *MemP2PMessenger) BroadcastOnChannel(channel string, topic string, buff []byte) {
+	fmt.Printf("memp2p:BroadcastOnChannel\n")
+	messenger.parametricBroadcast(topic, buff, false)
+}
+
+// Broadcast asynchronously sends the message to all peers in the network. It
+// calls parametricBroadcast() with async=true, which means that peers will
+// have their ReceiveMessage() function independently called as go-routines.
+func (messenger *MemP2PMessenger) Broadcast(topic string, buff []byte) {
+	fmt.Printf("memp2p:Broadcast\n")
+	messenger.parametricBroadcast(topic, buff, true)
+}
+
+// parametricBroadcast sends a message to all peers in the network, with the
+// possibility to choose from asynchronous or synchronous sending.
+func (messenger *MemP2PMessenger) parametricBroadcast(topic string, data []byte, async bool) {
+	if messenger.IsConnectedToNetwork() {
+		message, _ := NewMemP2PMessage(topic, data, messenger.ID())
+		for _, peer := range messenger.Network.Peers() {
+			if async == true {
+				go peer.ReceiveMessage(topic, message)
```

Comment:
Unhandled error.

Reply:
Added handling for the potential error returned by `NewMemP2PMessage`.

Reply:
Fixed with `go func() {  err := peer.ReceiveMessage(...)  }` which also logs the error. Otherwise, there should have been as many channels as there are peers to send the message to (according to @iulianpascalau).